File: 백준/Silver/11659. 구간 합 구하기 4/구간 합 구하기 4.js
let fs = require('fs');
let input = fs.readFileSync(0, 'utf8').trim().split('\n');

const [N, M] = input[0].split(' ').map((v) => +v);
const nums = input[1].split(' ').map((v) => +v);
const sections = input.slice(2).map((s) => s.split(' ').map((v) => +v));

function solution(N, M, nums, sections) {
  const result = [];
  const dp = new Array(N);
  dp[0] = nums[0];

  for (let i = 1; i < N; i++) {
    dp[i] = dp[i - 1] + nums[i];
  }

  for (let i = 0; i < M; i++) {
    const [start, end] = sections[i];
    if (start === 1) {
      result.push(dp[end - 1]);
    } else {
      result.push(dp[end - 1] - dp[start - 2]);
    }
  }

  return result.join('\n').trim()
}

console.log(solution(N, M, nums, sections));

File: 백준/Silver/11659. 구간 합 구하기 4/구간 합 구하기 4.py
[N, M] = map(int, input().split())
perm = list(map(int, input().split()))
ranges = []

for i in range(M):
    ranges.append(list(map(int, input().split())))


def solution(N, M, perm, ranges):
    psum = [0 for _ in range(N + 1)]
    for i in range(1, N + 1):
        psum[i] = psum[i - 1] + perm[i - 1]

    result = []
    for r in ranges:
        [start, end] = r
        result.append(psum[end] - psum[start - 1])

    return "\n".join(map(str, result))


print(solution(N, M, perm, ranges))
